### ISSUE TITLE
F# intellisense sometimes stops working 

### ISSUE THREAD

**Issue by daniel-wojcik** [dotnet/fsharp#12982]
I noticed that sometimes Intellisense in F# editor is not working (can't find any pattern): there are no member prompts, no syntax coloring and F12 key is also not functional. After spending some time to find the solution I came to conclusion that it is enough to delete **project.assest.json** file from obj project folder. After a few seconds after deleting this file all Intellisense functionality is restored.

Windows 10 21H2
VS 2022 17.2.0. Preview 2.0 and 3.0

**Comment by kerams** [comment]
I'm seeing something similar with FSharp.sln. It seems like I need to run Build.cmd every time before I open VS, otherwise the F# integration is completely dysfunctional.

**Comment by vzarytovskii** [comment]
I've checked the latest 17.2 preview as well as 17.3 from `main`, and don't see it anymore. I will close it for time being.

**Comment by daniel-wojcik** [comment]
@vzarytovskii I simply don't understand why this issue was closed - have you in any way verifed it? I'm constantly experiencing this problem, also on newest VS 17.2.0 Preview 6. Below two merged screenshots: upper before deleting obj/project.asset.json, lower: after deleting.

<img width="179" alt="FS_Intellisense_Problem" src="https://user-images.githubusercontent.com/8803692/166661222-64331bb6-64b9-4cab-a0c0-3e4756e0f48a.png">

**Comment by vzarytovskii** [comment]
> @vzarytovskii I simply don't understand why this issue was closed - have you in any way verifed it? I'm constantly experiencing this problem, also on newest VS 17.2.0 Preview 6. Below two merged screenshots: upper before deleting obj/project.asset.json, lower: after deleting.
> 
> <img width="179" alt="FS_Intellisense_Problem" src="https://user-images.githubusercontent.com/8803692/166661222-64331bb6-64b9-4cab-a0c0-3e4756e0f48a.png">
>  

I explained why I closed it - I was unable to reproduce it on the latest 17.2 and 17.3. on VisualFsharp solution with which I work daily. 

I will reopen it since apparently, it needs more investigation. The problem is to get it to reproduce so we can debug it.

**Comment by kerams** [comment]
VisualFSharp.sln is still broken for me _every time_ (it feels like) I open it without doing a restore with Build.cmd.

**Comment by vzarytovskii** [comment]
> VisualFSharp.sln is still broken for me _every time_ (it feels like) I open it without doing a restore with Build.cmd.

This is the problem, I don't see it, I have it currently open and everything seems to be fine (I'm on the latest 17.2 preview).

Do you not run build.cmd whatsoever?

**Comment by vzarytovskii** [comment]
Building it for the first time with build.cmd is required now 
cc @KevinRansom

**Comment by kerams** [comment]
> Building it for the first time with build.cmd is required for tooling to work properly, unfortunately.

I know, that's not the problem.

My routine:

- open the solution, see that nothing works, close it
- run Build.cmd and once the restore is complete, ctrl-c (I used to let the whole build finish, but it's not necessary - it seems the restore bit is important, which aligns with @daniel-wojcik's findings)
- open the solution again, integration works
- work on some autocompletion fixes, build the VS integration in VS and debug it in an experimental instance
- finish, close the solution, go to bed
- the next day go to step 1

**Comment by vzarytovskii** [comment]
> > Building it for the first time with build.cmd is required for tooling to work properly, unfortunately.
> 
> The first time after cloning/clean, I know, that's not the problem.
> 
> My routine:
> 
> - open the solution, see that nothing works, close it
> - run Build.cmd and once the restore is complete, ctrl-c (I used to let the whole build finish, but it's not necessary - it seems the restore bit is important, which aligns with @daniel-wojcik's findings)
> - open the solution again, integration works
> - work on some autocompletion fixes, build the VS integration in VS and debug it in an experimental instance
> - finish, close the solution, go to bed
> - the next day go to step 1

Hm, I see. I will try to repro with these steps.

Build.cmd is also needed for building the Proto compiler. 

Can you try opening compiler service solution and see if anything is different for you? It's in the `service` directory.

**Comment by kerams** [comment]
![explorer_6wA7fNB5fj](https://user-images.githubusercontent.com/5063478/166664672-5917bccf-e23a-418a-980a-f86649b133ae.png)

`Failed to restore FSharp.Test.Utilities` flashed by in the status bar.

**Comment by kerams** [comment]
Interestingly build.cmd does not help with this solution. When I try to do a build in VS, I get `C:\Program Files\dotnet\sdk\6.0.300-preview.22204.3\Sdks\Microsoft.DotNet.Arcade.Sdk\Sdk not found. Check that a recent enough .NET SDK is installed and/or increase the version specified in global.json.	
`

**Comment by vzarytovskii** [comment]
Before I started digging to deep @cartermp, sorry for pinging, but do you maybe know where should I start looking, or maybe this rings a bell to you?

**Comment by cartermp** [comment]
I wouldn't have a clue, but given there are two indicators (delete project.assets.json and the "failed to restore" message) it's likely a nuget restore failure. I don't know how to debug these though. I'd suggest pulling in someone from the project system or nuget client team to help figure it out.

**Comment by daniel-wojcik** [comment]
@vzarytovskii unfortunatelly I have no such solution which could be shared, but even on this one I'm using the problem occurs from time to time - no repro steps.

**Comment by kerams** [comment]
This has somehow gotten even worse in 17.3 Preview 1. I'm now regularly seeing the same problems in projects outside of VisualFSharp.sln almost without fail.

`fake build -t run`, which does a Paket restore, sometimes helps (some projects occasionally start working). The only remedy that appears to be reliable is deleting the `.vs` folder while VS is closed.

**Comment by vzarytovskii** [comment]
> This has somehow gotten even worse in 17.3 Preview 1. I'm now regularly seeing the same problems in projects outside of VisualFSharp.sln almost without fail.
> 
> `fake build -t run`, which does a Paket restore, sometimes helps (some projects occasionally start working). The only remedy that appears to be reliable is deleting the `.vs` folder while VS is closed.

I'm looking at it now, do you maybe have some public repo in mind you've seen it reproducing?

**Comment by kerams** [comment]
Sure, I've opened https://github.com/Zaid-Ajaj/Fable.Remoting at random (for the first time in a long time).

I had Fable.Remoting.DotnetClient/Proxy.fs and Fable.Remoting/MsgPack/Write.fs opened from last time. Proxy.fs was the active tab and after a while I started getting tooltips, but proper syntax coloring did not kick in until I reopened the tab. Write.fs was OK after I switched to it.

Then I reopened VS and the active Proxy.fs loaded up normally. Then I switched to Write.fs and nothing was working when the background tasks had finished. Reopening the tab did not help at all and other files in the MsgPack project behaved exactly the same with F# integration completely busted.

**Comment by vzarytovskii** [comment]
Ok, let me try and debug the completion provider and language service to see if I can have a repro locally.

**Comment by vzarytovskii** [comment]
Oh boy...

Latest `Int Preview`, just by installing the `vsix`, release version, VS is **not** under debug:
1. Don't pre-build `Fable-Remoting` (only have tools restored).
2. Open `Fable.Remoting.sln`, packages were restored.
3. Open `Fable.Remoting.DotnetClient/Proxy.fs` via picker (Ctrl-T).
4. Colouring kicked in straight away.
5. IntelliSense did show up on external `JsonConvert` symbol, as well as internal `FableJsonConverter`
6. Navigation did work too.
7. Open `Fable.Remoting/MsgPack/Write.fs` (different project).
8. Only basic colouring kicks in.
9. IntelliSense did **not** show up on both internal and external symbols. Even if triggered manually.
10. Navigation did **not** work.
11. Verify steps 7-10 in different projects, same behaviour.
12. Close and re-open solution, now it acts all-the-way around, for example, last document which was open is `Fable.Remoting/MsgPack/Write.fs`, the colouring and intellisence works there, but when I switch back to `Fable.Remoting.DotnetClient/Proxy.fs` (which worked before), I don't get colouring or intellisense, and it seems to **not**kick in if rebuilt.

**Same behaviour with cleaned up artifacts, but with pre-built `Fable-Remoting` using `build.cmd` script**:

Okay, it seems I can work with it.

FYI @dsyme @KevinRansom @KathleenDollard @TIHan 
It sounds like it may be related to BuildGraph, or FSharpDocument changes from https://github.com/dotnet/fsharp/pull/12305

**Comment by kerams** [comment]
So I am not completely loco, good to have a repro. The workaround of deleting `.vs` can be narrowed down to closing all tabs before exiting VS.

**Comment by vzarytovskii** [comment]
> So I am not completely loco, good to have a repro. The workaround of deleting `.vs` can be narrowed down to closing all tabs before exiting VS.

I did not try, but probably yes. And it still has some issues working out x-project. Trying to figure out why.

**Comment by auduchinok** [comment]
> Latest Int Preview, just by installing the vsix, release version, VS is not under debug:

To try to make it easier to find the cause, I've checked the repro in Rider with an FCS fork based on 1490dd5d8abfc5fe6609e56a174914cf72306179. The `Write.fs` file has all the features working instantly on the step 7.

So, it's either something that came after the commit we've based our FCS fork on or it's something Visual Studio specific.

**Comment by auduchinok** [comment]
There's a really small chance it could be fixed by https://github.com/dotnet/fsharp/pull/12570 which we have already merged in in our fork. Otherwise, we don't seem to have any related changes to the upstream.

**Comment by vzarytovskii** [comment]
> There's a really small chance it could be fixed by #12570 which is already merged in our fork. Otherwise, we don't seem to have any related changes to the upstream.

We did merge it to 17.3, I think.

**Comment by vzarytovskii** [comment]
**NOTE:** I have a laptop and a (significantly more powerful) desktop, and I **can** reproduce the bug 100% of the times on laptop, and **wasn't able** to reproduce it on desktop computer so far...

**Comment by vzarytovskii** [comment]
Apparently - more accurate repro is if document is opened **before** some background tasks completed, then intellisense/colouring/breadcrumbs are only shown in the project where this document belongs, and if we switch to another project, it's not getting checked (no results). This would explain why I couldn't reproduce it on my desktop PC, tasks were probably finished before I've opened document.

@auduchinok so, I think https://github.com/dotnet/fsharp/pull/12570 may actually help in this case. 
It's already merged into latest 17.3, and will try to reproduce it on it.

**Comment by vzarytovskii** [comment]
On the latest `main`, it works much better, typecheck/highlighting kicks in every time document is opened.
**However**, if the document is opened before (some) background tasks are complete, highlighting, breadcrumbs and typecheck does not kick in until reopened.

**Comment by chtoucas** [comment]
I experience the same problem since I installed VS 17.2 (through 17.2.2) on two different machines. I've tried deleting various stuff (.vs, bin, obj, packages.lock.json), opening an F# file only after all background tasks are complete,  but so far nothing works. The only thing that works _sometimes_ for me is unloading the project then reloading it.
Notice that the project I'm working on is a mix of C# and F# (for testing only); see the only pinned project on my profile.

**Comment by vzarytovskii** [comment]
> I experience the same problem since I installed VS 17.2 (through 17.2.2) on two different machines. I've tried deleting various stuff (.vs, bin, obj, packages.lock.json), opening an F# file only after all background tasks are complete,  but so far nothing works. The only thing that works _sometimes_ for me is unloading the project then reloading it.
> Notice that the project I'm working on is a mix of C# and F# (for testing only); see the only pinned project on my profile.
> 

Yeah, we're trying to get to the bottom of it. Experience is better in 17.3, but still getting some reports.

**Comment by marcin-krystianc** [comment]
I reproduced this issue in VS 2022 v17.3.0 Preview 1.1 with a very simple solution.

Steps to reproduce:
- Open [SmallFSharpSolution.zip](https://github.com/dotnet/fsharp/files/8897855/SmallFSharpSolution.zip) solution, open `Project001\MyModule.fs` and wait for IntelliSense.
- Close the solution, but do not close the Visual Studio.
- Open `SmallFSharpSolution.sln` solution again.
- Intellisense doesn't work anymore.

**Comment by vzarytovskii** [comment]
> I reproduced this issue in VS 2022 v17.3.0 Preview 1.1 with a very simple solution.
> 
> Steps to reproduce:
> 
> * Open [SmallFSharpSolution.zip](https://github.com/dotnet/fsharp/files/8897855/SmallFSharpSolution.zip) solution, open `Project001\MyModule.fs` and wait for IntelliSense.
> * Close the solution, but do not close the Visual Studio.
> * Open `SmallFSharpSolution.sln` solution again.
> * Intellisense doesn't work anymore.

Thanks, that's helpful, I will verify it later today. If it works (well, rather doesn't work) most of the times, it will be a huge help in troubleshooting it.

**Comment by vzarytovskii** [comment]
> No change towards the better in Preview 2. As far as I am concerned, this issue is having a critical impact (restarting VS multiple times to be able to work on different projects in the solution).

We are working other teams to fix it, sorry it is taking too long, trying to make sure we have proper fix for it.

**Comment by kerams** [comment]
No worries, I'm using a preview version, so I shouldn't complain too much. If 17.3 GA goes out like this though, it will be a catastrophe.

**Comment by vzarytovskii** [comment]
Ok, so, apparently, project system here:
https://github.com/dotnet/project-system/blob/2d8184bf78a11568ae502597ed5ba6d24bdd736c/src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/FSharp/FSharpCommandLineParserService.cs#L83-L94

Invokes our handler with empty sources, not sure whether it's a "diff" and by design, or not.

**Comment by vzarytovskii** [comment]
I've tested and merged the workaround from #13313 to the dev17.3.
They fix the ProjectSystem issue (I've verified it on the particular repro posted by @marcin-krystianc).

If someone can/wants to test it, you can install the VSIX - [VisualFSharpFull.zip](https://github.com/dotnet/fsharp/files/8928211/VisualFSharpFull.zip)

Meanwhile I will create an issue for the project-system to fix it there.

**Comment by kerams** [comment]
Looks OK, thanks.

**Comment by charlesroddie** [comment]
Great that this is fixed. Is 17.3 imminent? If there is a 17.2.x after the current one may I suggest this gets put in that. F# in VS is unusable without the fix.

> Servicing Updates are releases of targeted fixes for critical quality, performance, reliability, or security issues.

This is one of these!

**Comment by vzarytovskii** [comment]
> Great that this is fixed. Is 17.3 imminent? If there is a 17.2.x after the current one may I suggest this gets put in that. F# in VS is unusable without the fix.
> 
> > Servicing Updates are releases of targeted fixes for critical quality, performance, reliability, or security issues.
> 
> This is one of these!

It will unlikely make it to 17.2. It has been released already. I can suggest installing vsix for 17.2 as a workaround.

**Comment by charlesroddie** [comment]
> I can suggest installing vsix for 17.2 as a workaround.

This will work in my org, but it may be impossible to communicate this to all F# users. Maybe temporary notifications at the top of the main forums would work (slack, this repo, fsharp community forum), linking to the comment https://github.com/dotnet/fsharp/issues/12982#issuecomment-1158903384 . @ReedCopsey @dsyme is that reasonable?

I believe a large proportion of F# users will just conclude that it the language doesn't work, until VS 17.3 is out.

**Comment by rfreytag** [comment]
StackOverflow questions/answers get good search ranking so the question I posted "[[Visual Studio F# Intellisense stopped working](https://stackoverflow.com/questions/72741952/visual-studio-f-intellisense-stopped-working)](https://stackoverflow.com/questions/72741952/visual-studio-f-intellisense-stopped-working/72773018#72773018)" pointing back to this GitHub discussion #12982 may help get the word out.

If you have a better answer than the answer I cobbled together then please post and I will up vote.  

I of course welcome your up votes too.  :-)

**Comment by kerams** [comment]
> I'm not sure there will be another servicing release

If F# needs a critical fix, which this one arguably is, will the VS release team tell you tough luck? Do we only get to piggyback on someone elses's hotfix? Sad!

**Comment by rfreytag** [comment]
@kerams I had this problem and it was critical.  But then I restarted VS and recompiled and the Intellisense came back.  Does this not work for you or other?

See:   https://stackoverflow.com/questions/72741952/visual-studio-f-intellisense-stopped-working

**Comment by AlgorithmsAreCool** [comment]
@vzarytovskii To be clear, I am referring to the situation before your fix. (Thank you very much for your efforts btw!)

I assume this fix will remediate my issues with VS

**Comment by kerams** [comment]
@vzarytovskii, great news, thanks.

@rfreytag, it's a workaround, but an annoying one. You really don't want to do that every day (sometimes multiple times) for 2 months.

**Comment by vzarytovskii** [comment]
Visual Studio 2022 version 17.3 Preview 3 and Visual Studio 2022 version 17.2.6 were released, and include a fix.

**Comment by mauriciomagni** [comment]
Hello Everybody
I am experiencing this problem in version 17.4.3 of Visual Studio. I think it started in version 17.4.1 or 17.4.0.
My solution builds without errors and/or warnings and executes as expected. But Intellisense does not work.
I have tested the SmallFSharpSolution and it works as expected.
I have tried moving from net6 to net7, removing .vs folder, cleaning and rebuilding the solution, repairing Visual Studio, doing a partial load of the solution with just the needed project, but nothing has solve the problem.
I have noticed that for the first files in the solution intellisense works, but not for the last ones.
One more thing, in version 17.4.2, it show a Object null reference error signaling the m in module, at the beginning of a file where intellisense did not work, I cannot find that in version 17.4.3.

**Comment by mauriciomagni** [comment]
I have tested with 17.5.0 Preview 2.0, sadly, the error remains.

**Comment by T-Gro** [comment]
(issue resolved, was related to double listing of a file within the same .fsproj. Checking how best to report it and if there could be any legit situation of intentionally wanting to compile the same file twice)

**Comment by daniellittledev** [comment]
I'm running into this issue when using fsi files and a project reference (fsi files are in the referenced project). In the upstream project syntax highlighting stops working and when trying to inspect type information an error is shown.

`internal error: Object is not set to an instance of an object.`

```
Microsoft Visual Studio Professional 2022
Version 17.4.5
VisualStudio.17.Release/17.4.5+33403.182
Microsoft .NET Framework
Version 4.8.09032

Visual F# Tools   17.4.0-beta.22512.4+525d5109e389341bb90b144c24e2ad1ceec91e7b
Microsoft Visual F# Tools
```

![image](https://user-images.githubusercontent.com/1761789/220569689-773f0145-793f-4c1e-96ce-b39d45bde3cc.png)

**Comment by vzarytovskii** [comment]
> I'm running into this issue when using fsi files and a project reference (fsi files are in the referenced project). In the upstream project syntax highlighting stops working and when trying to inspect type information an error is shown.
> 
> `internal error: Object is not set to an instance of an object.`
> 
> ```
> Microsoft Visual Studio Professional 2022
> Version 17.4.5
> VisualStudio.17.Release/17.4.5+33403.182
> Microsoft .NET Framework
> Version 4.8.09032
> 
> Visual F# Tools   17.4.0-beta.22512.4+525d5109e389341bb90b144c24e2ad1ceec91e7b
> Microsoft Visual F# Tools
> ```
> 
> ![image](https://user-images.githubusercontent.com/1761789/220569689-773f0145-793f-4c1e-96ce-b39d45bde3cc.png)
> 

17.4.5 didn't get a fix, please try on 17.5

**Comment by daniellittledev** [comment]
@vzarytovskii are we talking VS 17.5 or F# 17.5, I've checked updated, is this a preview? how would I install it?

**Comment by vzarytovskii** [comment]
> @vzarytovskii are we talking VS 17.5 or F# 17.5, I've checked VS updates, is this a preview? how would I install it?

VS, should be able to download normally here
https://visualstudio.microsoft.com/downloads/

**Comment by daniellittledev** [comment]
Awesome, looks like it was released yesterday https://learn.microsoft.com/en-au/visualstudio/releases/2022/release-notes
